### DISCUSSION TITLE
problem with 'null' values

### DISCUSSION THREAD

**Discussion by scalpel4k** [stephenberry/glaze#1450] [General]
I have the following code

```
struct A{
 std::optional<int> i;
};

struct B {
 A a;
};

auto toString() {
  B b;
  return glz::write<glz::opts{
        .format = glz::JSON, .skip_null_members = 1, .prettify = 1, .validate_skipped = 0}>(b);
}
```
`struct A` has a custom serializer that extracts the value of `A::i`, so the result should look like this

```
{ "a" : 1 }
```
So far so good, this is working. But when `A::i` has no value I want the entry to be skipped, so I'm using `.skip_null_members = 1` as an option to the write method. However, I always end up getting an entry with a 'null' value.

I know the problem comes from the combination of those 2 objects. Is there a way to eliminate empty values right in the custom serializers?

thx Michi

**Comment by scalpel4k** [comment]
@stephenberry I've tried to understand the problem a bit better. When running my code in question I can see that the serializer calls the custom writer from the branch where `object_info.maybe_skipped` is `false` (write.hpp line 1570, version 4.0.1).

I can't (want to) make my wrapper optional because it would make the API of my data classes quite clunky.

Instead I was thinking if it would be possible to introduce a setting in the glz::meta<> where the user could specify some of the `glz::opts` like `maybe_skipped` or `error_on_unknown_keys`

Do you thinks that this could be a solution for the problem?

thx Michi

**Reply by stephenberry** [reply]
The problem is that Glaze doesn't have a means of knowing whether your object was serialized as null when you have custom serialization, because it might actually write something when the internal value is null.

I would first see if a custom serializer is needed for the `A` type. Can you explain some more about what this custom serializer does and why it is needed?

You probably could use glz::custom on `B` with some introspection on `A`, but I'd need to know more about the `A` type that you're actually working with.

**Reply by scalpel4k** [reply]
I fully understand your point that glaze does not know anything about the object, but I as a developer do. That's why I was asking whether I could maybe instrument the class in such a way that I could tell glaze to treat my `struct A` like it was a `std::optional`.

`A` in real life is actually a class template that encapsulates some data type on which it applies some validation before/after serialization in a generic way, plus it makes integral types distinct types that can only be applied in an appropriate context.

`struct A` also has methods/operators that make working with the wrapper as if it was the native data type. So if I would wrap my `struct A` in a `std::optional` I would have to refactor quite a bit of our application.

Currently we're still using nlohmann under the hood and I wanted to replace only the JSON serialization part without having to change anything in our other code base.

I indeed tried to use `glz::custom` but couldn't see how I can read back the data from the JSON string. With the serialization helper, however, I can provide the next "stage" a `std::string` (or whatever), let glaze do its magic, and then continue with my stuff.

Here's my custom reader w/o error handling:

```
template <KvpSchema S>
struct from<JSON, KVP<S>>
{
    template <auto Opts>
    static void op(KVP<S> &kvp, auto &&...args) noexcept(false) {
        read<JSON>::op<Opts>(kvp.value(), args...);
        kvp.validate();
    }
};
```

Does this answer your questions?

**Reply by stephenberry** [reply]
Thanks for the additional details, it's very helpful.

I think the easiest approach would be to get your type to satisfy Glaze's `nullable_t` concept:
```c++
template <class T>
concept nullable_t = !meta_value_t<T> && !str_t<T> && requires(T t) {
   bool(t);
   {
      *t
   };
};
```

Can you add an `operator bool()` to your class and an `operator*()`? If it had these things then Glaze would treat your class like a `std::optional` and `skip_null_members` should work.

**Reply by scalpel4k** [reply]
cool, that should work!

thanks a lot

**Reply by scalpel4k** [reply]
The devil's in the details ..... one of the data types is `bool` and I now get method redefinition errors on those types .....

**Reply by stephenberry** [reply]
Hmm, care to explain the method redefinition a bit more?

**Reply by scalpel4k** [reply]
A minimum example could look like this

```
template<KvpSchema S>
struct KVP {
  operator typename S::type_t() {
    return val.value();
  }

  operator bool() {
    return val.has_value();
  }

  std::optional<S::type_t> val;
};
```

`S::type_t` can be all sorts of things, also `bool` ....

**Reply by stephenberry** [reply]
Oh, interesting. What function do you normally use to check for a value, something like `has_value()`? We could potentially make a concept for a different API than a typical nullable type in C++.

**Reply by scalpel4k** [reply]
now we're talking ;-)

Personally I think that `glz::meta` would be the perfect place to specify things like that.

What do you think about this?

```
template<>
struct meta<T>
{
    static constexpr auto maybe_skipped = true;
};
```

Of course, making a concept that requires `has_value()` could also be used .... And yes, I currently have a `has_value()` method.

**Reply by stephenberry** [reply]
The `glz::meta` shouldn't be necessary. If we can just determine that the custom type is nullable, how to check for nullability, and how to get the value, then we should be good.

I think this pull request adds what you need: https://github.com/stephenberry/glaze/pull/1460

There are some basic unit tests to show similar behavior to what you have. Let me know your thoughts. You could try using this branch on your code.
